### PR TITLE
Add minimal example

### LINKED ISSUE #2057: Write minimal example
From slack discussion:

> David Irvine  2 minutes ago
I would go for absolute minimum. Add (and remove a few) nodes till a section splits. No need to send data around as there is enough data getting sections built up.

### PR DESCRIPTION
**Pull request by madadam**
Closes #2057

### CODE REVIEW

File: examples/minimal.rs [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+fn main() {
+    if cfg!(feature = "mock_base") {
+        panic!("This example must be built without the `mock_base` feature");
+    }
+
+    let opts = Options::from_args();
+    init_log(opts.verbosity);
+
+    if opts.count <= 1 {
+        start_single_node(opts.first, opts.bootstrap_contacts, opts.ip, opts.port)
```
[the remark this answers is not in the document]
Comment (madadam) [reply examples/minimal.rs]: You might want to start multiple instances of this example even across multiple machines, so the ability to start a single node is useful. Also "first node" isn't just the first node started in one instance of the example. It means the first node of the network. So you might want to start first node on one machine and normal node on another. That gives us the following options, all of which I think are valid:
1. Single first node
2. Single normal node
3. Multiple nodes where exactly one of them is first
4. Multiple nodes, all normal

File: examples/minimal.rs [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+// If `first` is true, the first spawned node will start as the first (genesis) node. The rest will
+// always start as regular nodes.
+fn start_multiple_nodes(
+    count: usize,
+    first: bool,
+    contacts: Vec<SocketAddr>,
+    ip: Option<IpAddr>,
+    base_port: Option<u16>,
+) {
+    let mut join_handles = Vec::with_capacity(count);
```
[the remark this answers is not in the document]
Comment (madadam) [reply examples/minimal.rs]: They join the threads when dropped (see the `ScopedJoinHandle` wrapper at the bottom of the file). But maybe we don't really need that?

File: examples/minimal.rs [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+        }
+    }
+
+    true
+}
+
+const TARGET_SELF: &str = "minimal";
+const TARGET_ROUTING: &str = "routing";
+
+fn init_log(verbosity: u8) {
```
Comment (dirvine) [inline examples/minimal.rs]: Would https://github.com/rust-cli/clap-log-flag perhaps make it easier? Our CLI etc. uses this so it owudl be more consistent?

File: src/chain/chain.rs [hunk 4]
```diff
@@ -218,7 +217,7 @@ impl Chain {
             Err(InsertError::ReplacedAlreadyInserted) => {
                 // TODO: If detecting duplicate vote from peer, penalise.
                 log_or_panic!(
```
Comment (dirvine) [inline src/chain/chain.rs]: I am not sure log_or_panic is valid anywhere except POC work, but less so here. As this is an example we can log errors or panic, but I don't think to mix them is good. In the cases where it seems to log is fine, unless we get into an unusable/UB state.

Comment (madadam) [reply src/chain/chain.rs]: We still use `log_or_panic` in a bunch of places and this PR isn't really about that. I'm OK with replacing `log_or_panic` with something better, but it should be done as a separate task. We also haven't fully decided what that better thing should be.

File: examples/minimal.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (maqi) [reply examples/minimal.rs]: maybe it's useful later on, for supporting exchanging messages.
However, I don't see it's necessary currently.

[the remark this answers is not in the document]
Comment (madadam) [reply examples/minimal.rs]: Actually I just tried it. We do need to collect the handles to keep the example running. Without it all the threads are killed immediately and the example terminates.

[the remark this answers is not in the document]
Comment (maqi) [reply examples/minimal.rs]: ok then

File: examples/minimal.rs (hunk 3 above)
Comment (madadam) [reply examples/minimal.rs]: Perhaps. But the logging logic I'm using currently is bit different. You can use `-v`, `-vv` and `-vvv` to enable logs (info, debug and trace respectively) from the routing crate only and `-vvvv` to enable logs from all crates. Doesn't seem I can achieve that with `clap-log-flag` which allows either only the current crate or all crates.

Also not sure such a simple feature justifies adding yet another dependency.

File: src/chain/chain.rs (hunk 4 above)
Comment (dirvine) [reply src/chain/chain.rs]: Yes, I agree, we will get to it soon enough. Thanks chap

**Pull request merged**